Support editing an override via the CLI

cverna opened the pull request:
This commit add support for editing overrides to the CLI.
By using the overrides edit command you can now expire an override using the
--expired flag.
Fixed #1049

Signed-off-by: Clement Verna <cverna@tutanota.com>

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -41,6 +41,19 @@
+    click.option('--user', envvar='USERNAME'),
+    click.option('--password', prompt=True, hide_input=True),
+    click.option('--staging', help='Use the staging bodhi instance',
+                 is_flag=True, default=False)]
```
If you add ```url_option``` to this list, you won't need to include ```@url_option``` on both the save and edit commands.

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -368,33 +381,44 @@ def query_buildroot_overrides(url, user=None, **kwargs):
     client = bindings.BodhiClient(base_url=url, username=user, password=password, staging=staging)
-    resp = client.save_override(nvr=nvr, duration=duration, notes=notes)
+    kwargs['edited'] = kwargs['nvr']
+    resp = client.save_override(**kwargs)
```
It seems that the only difference between save and edit is the ```kwargs['edited']``` line. What if we make a helper function that accepts a boolean called ```edit``` and then we can make the code just do

```
if edit:
    kwargs['edited'] = kwargs['nvr']
```

Then the save and edit function here can both share the other code.

bowlofeggs commented on bodhi/client/bindings.py:
```diff
@@ -273,7 +273,7 @@ def comment(self, update, comment, karma=0, email=None):
 
     @errorhandled
-    def save_override(self, nvr, duration, notes):
+    def save_override(self, **kwargs):
```
This change is not backwards compatible, and the bindings aren't only used by the Bodhi client. I think instead we should add a new ```edited``` parameter that defaults to ```False``` to preserve backwards compatibility, while continuing to allow ```nvr```, ```duration```, and ```notes``` to be positional arguments.

bowlofeggs commented on bodhi/tests/client/test___init__.py:
```diff
@@ -395,3 +394,40 @@ def test_url_flag(self, send_request, query):
+
+class TestEditBuilrootOverrides(unittest.TestCase):
+    """
+    Test the save_buildroot_overrides() function.
```
s/save/edit/

bowlofeggs commented on bodhi/client/bindings.py:
```diff
@@ -391,7 +387,7 @@ def testable(self):
 
     @staticmethod
-    def override_str(override):
+    def override_str(override, minimal=False):
```
To preserve backward compatibility, let's make ```minimal``` default to ```True```.

bowlofeggs commented on bodhi/client/bindings.py:
```diff
@@ -400,12 +396,22 @@ def override_str(override):
+        val += "  Notes: {}\n".format(override['notes'])
+        val += "  Expired: {}".format(override['expired_date'] is not None)
+
+        return val
```
This change seems fine in general, but it's also separate of adding editable buildroot overrides. Please break it into a second commit so each commit is atomic and specific in scope. You can include that commit in this same PR if you want to.

bowlofeggs commented on docs/man_pages/bodhi.rst:
```diff
@@ -90,6 +90,21 @@ The ``overrides`` command allows users to manage build overrides.
+        Notes on why this override is in place.
+
+    ``--expired``
+        Force an override to the expired state.
```
Since these options are shared with the save command, maybe there's a way to document them once and just mention that they are shared by the edit and save subcommands?

bowlofeggs reviewed (changes requested):
Cool, this looks great! Just a few things to change. Thanks!

cverna committed:
This commit add a detailed string reprensentation of an override.
A boolean is used to switch between minimal or detailed representation

Signed-off-by: Clement Verna <cverna@tutanota.com>

cverna committed:
This commit add support for editing overrides to the CLI.
By using the overrides edit command you can now expire an override using the
--expired flag.
Fixed #1049

Signed-off-by: Clement Verna <cverna@tutanota.com>

cverna commented:
@bowlofeggs Thanks for the review, I have followed your recommendations.

bowlofeggs commented:
Can you rebase this pull request to resolve the conflicts? Thanks!

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -17,11 +17,36 @@
 import subprocess
 import sys
 import traceback
-
```
This space is intentional - by PEP-8 imports should be grouped into three groups - Python built-ins, third party libraries, and then imports from your own project:

https://www.python.org/dev/peps/pep-0008/#imports

bowlofeggs replied:
Please add this space back to satisfy PEP-8.

bowlofeggs commented on bodhi/tests/client/test___init__.py:
```diff
@@ -450,3 +449,39 @@ def test_notes_and_notes_file(self):
 
             self.assertEqual(result.exit_code, 1)
             self.assertEqual(result.output, u'ERROR: Cannot specify --notes and --notes-file\n')
+
```
There should be two blank lines here for PEP-8. This is causing a test to fail:

```
/home/vagrant/bodhi/bodhi/tests/client/test___init__.py:453:1: E302 expected 2 blank lines, found 1
F.
======================================================================
FAIL: Enforce PEP-8 compliance on the codebase.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/bodhi/bodhi/tests/test_style.py", line 37, in test_code_with_flake8
    self.assertEqual(subprocess.call(flake8_command), 0)
AssertionError: 1 != 0
```

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -386,34 +423,38 @@ def query_buildroot_overrides(url, user=None, **kwargs):
+
+@overrides.command('edit')
+@add_options(save_edit_options)
+@click.option('--expired', help='Expire the override', is_flag=True, default=False)
```
Even though the API uses the past-tense for ```expired```, this isn't quite the best human interface because it's not an active verb. I think we should change this to ```--expire``` and then have the function convert the ```expire``` kwarg to ```expired``` for the API.

bowlofeggs commented on bodhi/client/bindings.py:
```diff
@@ -273,7 +273,7 @@ def comment(self, update, comment, karma=0, email=None):
+    def save_override(self, nvr, duration, notes, edit=False, expired=False):
         """ Save a buildroot override.
 
         This entails either creating a new buildroot override, or editing an
```
Let's add the new parameters to the docblock here.

bowlofeggs commented on bodhi/client/bindings.py:
```diff
@@ -287,14 +287,17 @@ def save_override(self, nvr, duration, notes):
                 'notes': notes,
                 'csrf_token': self.csrf(),
-            })
+                'duration': duration}
```
Does the API use ```duration```? I didn't find it after an admittedly brief search.

cverna replied:
Hum it does not but we have the `--duration` in the CLI , which is apparently not used.  I guess I will remove this and I can create an issue to implement the server side to set the duration using the CLI.

bowlofeggs replied:
The CLI ```--duration``` seems to be a convenience that is used to [calculate the ```expiration_date```](https://github.com/fedora-infra/bodhi/blob/2.5.0/bodhi/client/bindings.py#L288-L289) so I don't think we need to implement this server-side.

bowlofeggs commented on docs/release_notes.rst:
```diff
@@ -35,6 +35,8 @@ Features
   (`#1321 <https://github.com/fedora-infra/bodhi/issues/1321>`_).
 * The bodhi CLI now supports editing an update using the update id.
   (`#937 <https://github.com/fedora-infra/bodhi/issues/937>`_)
+* The bodhi CLI now supports editing an override using the nvr.
```
I think we can leave off the "using the nvr" because it makes it sound like it was possible to edit overrides some other way before (and thus minimizes how great this change is!)

bowlofeggs commented on bodhi/tests/client/test_bindings.py:
```diff
@@ -191,9 +191,9 @@ class TestBodhiClient_override_str(unittest.TestCase):
     """
```
Optional: It would be good to add two tests for the two new bindings API flag. We do have tests that hit that code through the CLI, but the bindings are their own interface that some third parties use too, so it would be good to test them independently. This is not required to merge.

cverna replied:
I would rather submit another PR for that.

bowlofeggs reviewed (approved):
This change looks good and seems to work well. Please add that one space back before merging!
